Add Grace's caputring data change in dbt DevBlog (#1673)
fishtown-analytics/docs.getdbt.com: website/blog/2022-07-12-change-data-capture-metrics.md

- **Downstream Incremental Model**: Build an incremental model downstream of your metrics model to “grab” every point-in-time version
- **Upstream Snapshots**: Build snapshots on all of your sources to capture changes in your raw data and calculate all versions of history every time you execute a `dbt run`

Finally, I’ll discuss the pros and cons of each solution to give you a head start on step two. 

## Scenario

Let’s return to Joanne. Using dbt and her favorite BI tool, Joanne has created an income report to track monthly revenue for each product category. 

You can imagine her DAG as shown below, where `fct_income` captures revenue metrics per month for each product category.

![](/img/blog/2022-07-12-change-data-capture-metrics/fct-income-dag.png)

Joanne executes a `dbt run` on January 30th, 2022 and queries the resulting table:

```sql
select * from fct_income where month_year = "January 2022"
```

She gets the following output:

| month_year | product_category | revenue | run_timestamp |
|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 100 | 01/30/22 12:00:00 | 2018-01-01 |
| January 2022 | electronics | 200 | 01/30/22 12:00:00 | 2018-01-01 |

> | January 2022 | books | 100 | 01/30/22 12:00:00 | 2018-01-01 |

Reviewer: ```suggestion
| month_year | product_category | revenue | run_timestamp |
|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 100 | 01/30/22 12:00:00 | 
| January 2022 | electronics | 200 | 01/30/22 12:00:00 | 
| January 2022 | books | 100 | 01/30/22 12:00:00 | 
```

But a few days later, her source data changes for January - a manufacturing cost was dated incorrectly, and now has been updated in the source. Joanne executes a `dbt run` again on February 3rd. Now when she queries `fct_income`, she gets the following output:

> ![](/img/blog/2022-07-12-change-data-capture-metrics/table-2.png)

Reviewer: ```suggestion
| month_year | product_category | revenue | run_timestamp |
|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 50 | 02/03/22 16:00:00 | 
| January 2022 | electronics | 150 | 02/03/22 16:00:00 | 
| January 2022 | books | 200 | 02/03/22 16:00:00 | 
```

Reviewer: Would be good to bold / italicize / change color of the changed revenue column

Will also need to delete the image file

A few days later, Joanne finds a bug in her `dbt code`. She fixes the bug and executes a dbt run again on February 10th. Now, when she queries `fct_income`, she gets the following output:

> ![](/img/blog/2022-07-12-change-data-capture-metrics/table-3.png)

Reviewer: ```suggestion
| month_year | product_category | revenue | run_timestamp |
|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 52 | 02/10/22 08:00:00 | 
| January 2022 | electronics | 152 | 02/10/22 08:00:00 | 
| January 2022 | books | 202 | 02/10/22 08:00:00 | 
```

Reviewer: Would be good to bold / italicize / change color of the changed revenue column

Will also need to delete the image file

```sql
{{
    config(
        materialized='incremental'
    )
}}
```

By materializing `int_income_history` as incremental but *not* including a `unique_key` config, dbt will only execute `INSERT` statements – new rows will be added, but old rows will remain unchanged.

The rest of `int_income_history` would look like this:

```sql
...

select
    *
from {{ ref('fct_income') }}
{% if is_incremental() %}
    where true
{% endif %}
```

There are a few additional configs that Joanne might find helpful:

- she can use the `on_schema_change` config to handle schema changes if  new columns are added and/or deleted from `fct_income`
- she can also set the `full_refresh` config to false in order to prevent accidental loss of the historical data 
- she can build this table in a custom `schema` if she wants to enforce specific role-based permissions for this historical table
- she can specify a time-grain `unique_key` if she wants to reduce the amount of versions being captured 

>     - for example, if she only wants to capture the final version of each day she could set `unique_key = date_trunc('day', run_timestamp)`. This is excluded from the example below, as we are making the assumption that Joanne does indeed want to capture every version of every version of `fct_income`.

Reviewer: ```suggestion
    - for example, if she only wants to capture the final version of each day she could set `unique_key = date_trunc('day', run_timestamp)`. This is excluded from the example below, as we are making the assumption that Joanne does indeed want to capture every version of `fct_income`
```

	  on_schema_change='sync_all_columns'
    )
}}
```

As a final step, Joanne would create `fct_income_history` to join in the seed file to determine which versions are “correct”. Her new DAG looks like this, where `int_income_history` is an incremental model without a unique key:

![](/img/blog/2022-07-12-change-data-capture-metrics/int-income-history-dag.png)

The final output of `fct_income_history` would look identical to `stg_snapshot_fct_income` from her initial approach: 

| month_year | product_category | revenue | run_timestamp | correct_version |
|:---:|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 100 | 01/30/22 12:00:00 | FALSE |
| January 2022 | electronics | 200 | 01/30/22 12:00:00 | FALSE |
| January 2022 | books | 300 | 01/30/22 12:00:00 | FALSE |
| January 2022 | clothing | 50 | 02/03/22 16:00:00 | FALSE |
| January 2022 | electronics | 150 | 02/03/22 16:00:00 | FALSE |
| January 2022 | books | 200 | 02/03/22 16:00:00 | FALSE |
| January 2022 | clothing | 52 | 02/10/22 08:00:00 | TRUE |
| January 2022 | electronics | 152 | 02/10/22 08:00:00 | TRUE |
| January 2022 | books | 202 | 02/10/22 08:00:00 | TRUE |

## Solution #2: Upstream Snapshots

> Alternatively, Joanne could snapshot her source data and add flexibility to her modeling so that all historical versions are calculated `at the same time`. Let’s look at our example.

Reviewer: ```suggestion
Alternatively, Joanne could snapshot her source data and add flexibility to her modeling so that all historical versions are calculated *at the same time*. Let’s look at our example.
```

Joanne could track changes in the source data by adding snapshots directly on top of her raw data.

![](/img/blog/2022-07-12-change-data-capture-metrics/snapshots-dag.png)

This would *change the <Term id="grain" />* of these `stg_` tables, so she would see a row for each version of each field. The staging models will contain the history of each record. 

Remember the source data change Joanne noticed — a manufacturing cost was dated incorrectly (Junkuary 2022 instead of January 2022). With this solution, the `costs_snapshot` model will pick up this change:

```sql
{% snapshot costs_snapshot %}

{{
    config(
      target_database='analytics',
      target_schema='snapshots',

>       unique_key='id',

Reviewer: ```suggestion
      unique_key='cost_id',
```

      strategy='timestamp',
      updated_at='updated_at',
    )
}}

select * from {{ source('source', 'costs') }}

{% endsnapshot %}
```

> | id | month_year | cost | updated_at | dbt_valid_from | dbt_valid_to |

Reviewer: ```suggestion
| cost_id | month_year | cost | updated_at | dbt_valid_from | dbt_valid_to |
```

> | id | month_year | cost | updated_at | dbt_valid_from | dbt_valid_to |
> |:---:|:---:|:---:|:---:|:---:|:---:|
> | 1 | Junkuary 2022 | 50 | 01/15/22 12:00:00 | 01/15/22 12:00:00 | 02/03/22 12:00:00 |
> | 1 | January 2022 | 50 | 02/03/22 12:00:00 | 02/03/22 12:00:00 | NULL |

Reviewer: ```suggestion
| id | month_year | cost | tax | updated_at | dbt_valid_from | dbt_valid_to |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | Junkuary 2022 | 50 | 1 | 01/15/22 12:00:00 | 01/15/22 12:00:00 | 02/03/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 02/03/22 12:00:00 | 02/03/22 12:00:00 | NULL |
```

:::note Note
Because snapshots only capture changes detected at the time the dbt snapshot command is executed, it is technically possible to miss some changes to your source data. You will have to consider how often you want to run this snapshot command in order to capture the history you need.
:::

> `fct_income_history` now calculates the metrics for each version of source data, every time Joanne executes a `dbt run`. In other words, the downstream `fct_` models are **version-aware**.

Reviewer: ```suggestion
The original `fct_income` model now calculates the metrics for each version of source data, every time Joanne executes a `dbt run`. In other words, the downstream `fct_` models are **version-aware**. Because of this, Joanne changes the name of `fct_income` to `fct_income_history` to be more descriptive.
```

In order to track changes in business logic, she can create case-statements in her models that depend on the valid date of a given field. 

```sql
case
   when dbt_valid_to < some_date
      then <original_logic>
   when dbt_valid_to between some_date and next_date
      then <new_logic>
   else 
       <current_logic> 
end as metric_x
```

*Note: This case-statement can even be pulled out into a macro if the number of changes becomes substantial.*

> 

Reviewer: ```suggestion

Remember the bug Joanne found in her dbt code. With this solution, the updated logic will be captured as a case statement in her `stg_costs` model:

```sql
case
   when dbt_valid_to < to_timestamp('02/10/22 08:00:00')
      then cost + tax
   else 
       cost 
end as final_cost

```

^ Can't figure out how to add the final ``` without messing up the preview

> In order to track changes in business logic, she can create case-statements in her models that depend on the valid date of a given field. 
> 
> ```sql
> case
>    when dbt_valid_to < some_date
>       then <original_logic>
>    when dbt_valid_to between some_date and next_date
>       then <new_logic>
>    else 
>        <current_logic> 
> end as metric_x
> ```
> 
> *Note: This case-statement can even be pulled out into a macro if the number of changes becomes substantial.*
> 
> After joining in the seed file to determine which versions are “correct” (check out [Tackling the complexity of joining snapshots](https://docs.getdbt.com/blog/joining-snapshot-complexity)), her new DAG looks like this:
> 
> ![](/img/blog/2022-07-12-change-data-capture-metrics/final-dag.png)
> 
> Again, the final output of `fct_income_history` would look identical to `stg_snapshot_fct_income` from her initial approach: 
> 
> | month_year | product_category | revenue | run_timestamp | correct_version |
> |:---:|:---:|:---:|:---:|:---:|
> | January 2022 | clothing | 100 | 01/30/22 12:00:00 | FALSE |
> | January 2022 | electronics | 200 | 01/30/22 12:00:00 | FALSE |
> | January 2022 | books | 300 | 01/30/22 12:00:00 | FALSE |
> | January 2022 | clothing | 50 | 02/03/22 16:00:00 | FALSE |
> | January 2022 | electronics | 150 | 02/03/22 16:00:00 | FALSE |
> | January 2022 | books | 200 | 02/03/22 16:00:00 | FALSE |
> | January 2022 | clothing | 52 | 02/10/22 08:00:00 | TRUE |
> | January 2022 | electronics | 152 | 02/10/22 08:00:00 | TRUE |
> | January 2022 | books | 202 | 02/10/22 08:00:00 | TRUE |

Reviewer: ```suggestion
In order to track changes in business logic, she can apply each version of logic to the relevant records and union together. 

Remember the bug Joanne found in her dbt code. With this solution, she can track this change in business logic in the `stg_costs` model:

```sql
-- apply the old logic for any records that were valid on or before the logic change
select 
	cost_id,
	...,
	cost + tax as final_cost, -- old logic
       1 || ‘-’ || dbt_valid_from as version
from costs_snapshot
where dbt_valid_from <= to_timestamp('02/10/22 08:00:00')

union all

-- apply the new logic for any records that were valid after the logic change
select 
	cost_id,
	...,
	cost as final_cost, -- new logic
       2 || ‘-’ || dbt_valid_from as version
from costs_snapshot
where to_timestamp('02/10/22 08:00:00') between dbt_valid_to and coalesce(dbt_valid_from, to_timestamp('01/01/99 00:00:00'))
```

| cost_id | month_year | cost | tax | final_cost | version |
|:---:|:---:|:---:|:---:|:---:| :---:|
| 1 | Junkuary 2022 | 50 | 1 | 51 | 1 - 01/15/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 51 | 1 - 02/03/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 50 | 1 - 02/03/22 12:00:00 |

The contents of the seed income_report_versions would look slightly different to match the change in version definition:
| month_year | correct_version | comment |
|:---:|:---:|:---:|
| January 2022 | 2 - 02/03/22 12:00:00 | Approved by Lucy |

After joining in the seed file (check out [Tackling the complexity of joining snapshots](https://docs.getdbt.com/blog/joining-snapshot-complexity)), her new DAG looks like this:

![](/img/blog/2022-07-12-change-data-capture-metrics/final-dag.png)

The final output of fct_income_history would accomplish the same goal as stg_snapshot_fct_income from her initial approach: 

| month_year | product_category | revenue | version | correct_version |
|:---:|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 100 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | electronics | 200 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | books | 300 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | clothing | 50 | 1 - 02/03/22 12:00:00 | FALSE |
| January 2022 | electronics | 150 | 1 - 02/03/22 12:00:00 | FALSE |
| January 2022 | books | 200 | 1 - 02/03/22 12:00:00 | FALSE | 
| January 2022 | clothing | 52 | 2 - 02/03/22 12:00:00 | TRUE |
| January 2022 | electronics | 152 | 2 - 02/03/22 12:00:00 | TRUE |
| January 2022 | books | 202 | 2 - 02/03/22 12:00:00 | TRUE |

```

## Final thoughts

Both of these solutions allow Joanne to achieve her desired output – a table containing all versions of income metrics for a given month – while improving the workflow and the efficiency of the final model.

However, each has its advantages and disadvantages. 

**Solution #1: Downstream Incremental Model**

| Pros | Cons |
|:---:|:---:|
| incremental models without unique keys are fast | this isn't really the intended use of the incremental <Term id="materialization" /> |
|  | Joanne has no way to re-calculate prior versions if her historical table is accidentally lost |

**Solution #2: Upstream Snapshots**

> 

Reviewer: Is there a way to make the columns in both Pro / Con tables the same size?

| Pros | Cons |
|:---:|:---:|
| Joanne doesn't have to worry about losing historical data | snapshots are highly complex and require more institutional knowledge for Joanne's team |

> |  | every time Joanne wants to make a code change that affects her metric calculations, she'll have to remember to write the change as a case/when statement |

Reviewer: ```suggestion
|  | every time Joanne wants to make a code change that affects her metric calculations, she'll have to remember to apply the change to each set of relevant records and union the outputs together |
```

The rest of `int_income_history` would look like this:

```sql
...

select
    *
from {{ ref('fct_income') }}
{% if is_incremental() %}
    where true
{% endif %}
```

There are a few additional configs that Joanne might find helpful:

- she can use the `on_schema_change` config to handle schema changes if  new columns are added and/or deleted from `fct_income`
- she can also set the `full_refresh` config to false in order to prevent accidental loss of the historical data 
- she can build this table in a custom `schema` if she wants to enforce specific role-based permissions for this historical table
- she can specify a time-grain `unique_key` if she wants to reduce the amount of versions being captured 
    - for example, if she only wants to capture the final version of each day she could set `unique_key = date_trunc('day', run_timestamp)`. This is excluded from the example below, as we are making the assumption that Joanne does indeed want to capture every version of `fct_income`

The final config block for `int_income_history` might look something like this:

```sql
{{
    config(

>         materialized='incremental',
> 	  full_refresh=false,
> 	  schema='history',
> 	  on_schema_change='sync_all_columns'

Reviewer: ```suggestion
        materialized='incremental',
	full_refresh=false,
	schema='history',
	on_schema_change='sync_all_columns'
```

|:---:|:---:|:---:|:---:|:---:|:---:|
| 1 | Junkuary 2022 | 50 | 01/15/22 12:00:00 | 01/15/22 12:00:00 | 02/03/22 12:00:00 |
| 1 | January 2022 | 50 | 02/03/22 12:00:00 | 02/03/22 12:00:00 | NULL |

:::note Note
Because snapshots only capture changes detected at the time the dbt snapshot command is executed, it is technically possible to miss some changes to your source data. You will have to consider how often you want to run this snapshot command in order to capture the history you need.
:::

The original `fct_income` model now calculates the metrics for each version of source data, every time Joanne executes a `dbt run`. In other words, the downstream `fct_` models are **version-aware**. Because of this, Joanne changes the name of `fct_income` to `fct_income_history` to be more descriptive.

In order to track changes in business logic, she can apply each version of logic to the relevant records and union together. 

Remember the bug Joanne found in her dbt code. With this solution, she can track this change in business logic in the `stg_costs` model:

```sql
-- apply the old logic for any records that were valid on or before the logic change
select 
	cost_id,
	...,

> 	cost + tax as final_cost, -- old logic
>        1 || ‘-’ || dbt_valid_from as version

Reviewer: ```suggestion
	cost + tax as final_cost, -- old logic
        1 || ‘-’ || dbt_valid_from as version
```

from costs_snapshot
where dbt_valid_from <= to_timestamp('02/10/22 08:00:00')

union all

-- apply the new logic for any records that were valid after the logic change
select 
	cost_id,
	...,

> 	cost as final_cost, -- new logic
>        2 || ‘-’ || dbt_valid_from as version

Reviewer: ```suggestion
	cost as final_cost, -- new logic
        2 || ‘-’ || dbt_valid_from as version
```

from costs_snapshot
where to_timestamp('02/10/22 08:00:00') between dbt_valid_to and coalesce(dbt_valid_from, to_timestamp('01/01/99 00:00:00'))
```

> 

Reviewer: ```suggestion
| cost_id | month_year | cost | tax | final_cost | version |
|:---:|:---:|:---:|:---:|:---:| :---:|
| 1 | Junkuary 2022 | 50 | 1 | 51 | 1 - 01/15/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 51 | 1 - 02/03/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 50 | 1 - 02/03/22 12:00:00 |

The contents of the seed `income_report_versions` would look slightly different to match the change in version definition:
| month_year | correct_version | comment |
|:---:|:---:|:---:|
| January 2022 | 2 - 02/03/22 12:00:00 | Approved by Lucy |

After joining in the seed file (check out [Tackling the complexity of joining snapshots](https://docs.getdbt.com/blog/joining-snapshot-complexity)), her new DAG looks like this:

![](/img/blog/2022-07-12-change-data-capture-metrics/final-dag.png)

The final output of `fct_income_history` would accomplish the same goal as `stg_snapshot_fct_income` from her initial approach: 

| month_year | product_category | revenue | version | correct_version |
|:---:|:---:|:---:|:---:|:---:|
| January 2022 | clothing | 100 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | electronics | 200 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | books | 300 | 1 - 01/15/22 12:00:00 | FALSE |
| January 2022 | clothing | 50 | 1 - 02/03/22 12:00:00 | FALSE |
| January 2022 | electronics | 150 | 1 - 02/03/22 12:00:00 | FALSE |
| January 2022 | books | 200 | 1 - 02/03/22 12:00:00 | FALSE | 
| January 2022 | clothing | 52 | 2 - 02/03/22 12:00:00 | TRUE |
| January 2022 | electronics | 152 | 2 - 02/03/22 12:00:00 | TRUE |
| January 2022 | books | 202 | 2 - 02/03/22 12:00:00 | TRUE |

```

Remember the bug Joanne found in her dbt code. With this solution, she can track this change in business logic in the `stg_costs` model:

```sql
-- apply the old logic for any records that were valid on or before the logic change
select 
	cost_id,
	...,
	cost + tax as final_cost, -- old logic
        1 || ‘-’ || dbt_valid_from as version
from costs_snapshot
where dbt_valid_from <= to_timestamp('02/10/22 08:00:00')

union all

-- apply the new logic for any records that were valid after the logic change
select 
	cost_id,
	...,
	cost as final_cost, -- new logic
        2 || ‘-’ || dbt_valid_from as version
from costs_snapshot
where to_timestamp('02/10/22 08:00:00') between dbt_valid_to and coalesce(dbt_valid_from, to_timestamp('01/01/99 00:00:00'))
```
| cost_id | month_year | cost | tax | final_cost | version |
|:---:|:---:|:---:|:---:|:---:| :---:|
| 1 | Junkuary 2022 | 50 | 1 | 51 | 1 - 01/15/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 51 | 1 - 02/03/22 12:00:00 |
| 1 | January 2022 | 50 | 1 | 50 | 1 - 02/03/22 12:00:00 |

The contents of the seed `income_report_versions` would look slightly different to match the change in version definition:

> | month_year | correct_version | comment |

Reviewer: ```suggestion

| month_year | correct_version | comment |
```
